File: java/src/main/java/problems201_300/ReverseLinkedList.java
package problems201_300;

import concepts.SinglyListNode;

enum ReverseLinkedList {
    DEFAULT {
        @Override
        SinglyListNode reverseList(SinglyListNode head) {
            SinglyListNode previous = null;
            SinglyListNode current = head;
            while (current != null) {
                SinglyListNode next = current.next;
                current.next = previous;
                previous = current;
                current = next;
            }
            return previous;
        }
    };

    abstract SinglyListNode reverseList(SinglyListNode head);
}

File: kotlin/src/main/kotlin/problems201_300/ReverseLinkedList.kt
package problems201_300

import concepts.SinglyListNode

enum class ReverseLinkedList {
    DEFAULT {
        override fun reverseList(head: SinglyListNode?): SinglyListNode? {
            var previous: SinglyListNode? = null
            var current = head
            while (current != null) {
                val next = current.next
                current.next = previous
                previous = current
                current = next
            }
            return previous
        }
    },
    ;

    abstract fun reverseList(head: SinglyListNode?): SinglyListNode?
}
